src/python/pants/backend/jvm/tasks/resources_task.py

```diff
@@ -49,18 +54,39 @@ def execute(self):
                           fingerprint_strategy=self.create_invalidation_strategy(),
                           invalidate_dependents=False,
                           topological_order=False) as invalidation:
-      for vt in invalidation.all_vts:
+      for vt in invalidation.invalid_vts:
+        # Generate resources to the chroot.
+        self.prepare_resources(vt.target, vt.results_dir)
+        processed_targets.append(vt.target)
+      for vt, digest in self._capture_resources(invalidation.all_vts):
         # Register the target's chroot in the products.
         for conf in self.get_options().confs:
-          runtime_classpath.add_for_target(vt.target, [(conf, vt.results_dir)])
-        # And if it was invalid, generate the resources to the chroot.
-        if not vt.valid:
-          self.prepare_resources(vt.target, vt.results_dir)
-          processed_targets.append(vt.target)
-          vt.update()
```

Comment:
It's called implicitly at the end of the block: while it would have been fine to leave this here, resource generation is not generally expensive, which is the reason to allow it to fail incrementally by calling `vt.update()` explicitly.